rfc-173-testing-our-work.md

```diff
@@ -0,0 +1,68 @@
+# Testing our work
+
+## Summary
+
+Testing the code we write, as we write it, is vital to delivering quality
+software in a sustainable way.
+
+Testing our work has always been a core development principle for dxw. Our RFC
+process was introduced relatively recently and there's no good reason why
+testing hasn't been added. This proposal aims to fill this gap and officially
+document our stance on testing as well as provide some sensible defaults.
+
+## Problem
+
+- Testing pattern divergence - the subject of writing tests is a broad one and
+  there are naturally varying opinions on what patterns to follow in any
+  situation. The patterns we use have been slowing diverging, we think this is
+  largely a natural side affect of different team compositions and their views
+  and lack of explicit standards to follow
+- Tooling divergence - we work with the philosophy and freedom of using the best
+  tool for the job. It is okay to diverge when there is good reason but we could
+  benefit from using like-for-like tools through defining some defaults. The
+  impact of using lots of different tooling is that projects are harder for us
+  to swap between and are harder for us to support
+- Team growth - the development team is growing and we often invite contractors
+  to represent us. Having an agreed set of principles will help all new joiners
+  build an understanding around what we expect from one another when writing
+  software
+- Stack variety - dxw are increasingly working with different stacks. These
+  often require us to use different testing tools. Having a way to record our
+  preferences as we grow will make it easier to contribute or work on those
+  stacks again in future
+
+## Proposal
+
+The topic of testing code has many facets. The aim of this proposal is to
+capture some of the higher level principles we wish to follow, seperate to how
+we achieve them. We propose to document those finer details in our
+[standards for testing](https://docs.google.com/document/d/1pSZkrviFAiR-0U-PGixOyC2MfiBdCIWIN66YyLa23nc/edit#heading=h.owf78h5xtccv)
+that may change more frequently. These standards may eventually become a guide
+in [the dxw Playbook](https://github.com/dxw/playbook).
+
+- We MUST write automated tests for all the code we write that will (or could)
+  serve people in live environments – this includes internal tools for dxw staff
+  and tooling that provisions real infrastructure
+- We MUST write appropriate test coverage for the code we write, inline with our
+  standards for testing. This includes using a combination of feature, request
+  and unit tests. It will also involve setting testing boundaries between
+  systems and third-party software
+- We MUST configure our continuous integration tools to automatically run as
+  deployment pipelines. It MUST prevent deployments completing should any test
+  fail
+- We MUST set up new projects to require 100% code coverage
```

Comment:
I think 100% code coverage is pretty much impossible in many cases. There will always be some lines of code that are going to be hard / impossible to test. I think 98% might be more realistic.

Reply:
I think in those cases we should explicitly remove them from the coverage with `# :nocov:` comments rather than allowing a couple of degrees of untestable padding?

Reply:
Yeah, that's fair. Might be worth codifying that in that case 👍